### DISCUSSION TITLE
Runtime extraction - doc are too confusing. Still have no clue...

### DISCUSSION THREAD

**Discussion by OnkelTem** [i18next/i18next#1741] [General]
I was going to start using Runtime Extraction for my react project with i18next, but after two hours of reading docs and exploring the suggested packages like `i18next-http-backend` I'm not any closer as I was in the beginning.

Here is the original link to Runtime extraction docs: https://www.i18next.com/how-to/extracting-translations#3-runtime-extraction

It says:

> In case of the http-backend just set `saveMissing: true` on init:

```js
import i18n from "i18next";
import detector from "i18next-browser-languagedetector";
import backend from "i18next-http-backend";
import { initReactI18next } from "react-i18next";

i18n
  .use(detector)
  .use(backend)
  .use(initReactI18next) // passes i18n down to react-i18next
  .init({
    fallbackLng: "en", // use en if detected lng is not available
    saveMissing: true // send not translated keys to endpoint
  });

export default i18n;
```

The snippet really looks like it's used in the front-end: same `i18n.init()`, same `i18next-browser-languagedetector`.
So I enabled it and then... what? Who is gonna serve the requests? I see no mention of it in the docs.

Also this:

> `import backend from "i18next-http-backend";`

Is it the backend itself or a plugin which is assumed to talk to _some_ backend?

Amazingly, but the whole project page here https://github.com/i18next/i18next-http-backend doesn't make it _any clearer_.
The provided options example mentions some URLs, some patterns, some queries, but what is it about - a front-end or a back-end or maybe some combination? Does it need a server or it is a server itself?

The `examples` doesn't help as well. Take the `node` example: https://github.com/i18next/i18next-http-backend/tree/master/example/node

The `app.js` file is definitely a server, but uses the syntax from the i18next client. Then maybe it should be used both in the server and in the client side?

This is very confusing folks. Docs shouldn't be that cryptic and unclear.

**Comment by jamuhl** [comment]
https://www.i18next.com/how-to/extracting-translations#3-runtime-extraction says it all:

---> Check the [options](https://github.com/i18next/i18next-xhr-backend#backend-options) for where missing translation gets sent. 


https://github.com/i18next/i18next-xhr-backend#backend-options so backend sends missing to the in the backend defined endpoint, which might  be:

- some customer server you control
- some customer express server using: https://github.com/i18next/i18next-http-middleware#add-routes
- some 3rd party translation management like locize.com

My guess is you expect saveMissing doing some magic write to your local filesystem during development...but that's not the way it works -> must be some real server with an endpoint processing those missing requests

**Comment by OnkelTem** [comment]
@jamuhl 

Thank you for the quick reply.

So while the first sentence from `i18next-http-backend` page says:

**This is a simple i18next backend to be used in Node.js...**

as I assumed `i18next-http-backend` is not a back-end despite its name. It's actually a client which talks to some back-end.

> some customer express server using: https://github.com/i18next/i18next-http-middleware#add-routes

So it's an express middleware which has some `missingKeyHandler()` handler. Which does something.

Ok, so there is no a solution for this. I thought that runtime extraction is used for dynamic stuff (when strings are dynamic) but it doesn't seem to be the case.

And yeah, 'saveMissing' - just another confusing key :) 

My goodness...

**Reply by jamuhl** [reply]
not sure what you mean with "dynamic stuff"

save missing keys using a backend plugin to your backend...idk...if you got better ideas for naming...or like to add PR's to make the documentation more user friendly...open for all help we can get
